Log Level Config

arthur740212 opened the pull request:
The log levels on startup used to be hard coded.
Now it is loaded from a config file to allow flexibility when debugging with logs.
![eos_plugin_for_unity - Lobbies - Windows, Mac, Linux - Unity 2021 3 8f1 _DX11_ 3_15_2024 4_07_36 PM](https://github.com/PlayEveryWare/eos_plugin_for_unity/assets/36757173/7840b0e4-206d-4d47-8fb4-f2ae517bb55f)
![eos_plugin_for_unity - Lobbies - Windows, Mac, Linux - Unity 2021 3 8f1 _DX11_ 3_15_2024 4_07_41 PM](https://github.com/PlayEveryWare/eos_plugin_for_unity/assets/36757173/e68767e5-8485-4a1b-aefd-fcccbce18a11)

arthur740212 committed:
feat(log) : startup log level config in string form

Unity json doesn't support enum nor dictionaries, thus a structure with two strings are used in the config instead

arthur740212 committed:
feat(log) : Helper for converting strings to enums

arthur740212 committed:
feat(log,editor) : editor window for log level config

arthur740212 committed:
feat(log) : example log level config file

arthur740212 committed:
feat(log) : use log config during appication startup/init

paulhazen commented on Assets/Plugins/Source/Core/EOSManager.cs:
```diff
@@ -588,19 +590,22 @@ public void Init(IEOSCoroutineOwner coroutineOwner, string configFileName)
-#endif
+
+                    for (int logCategoryIndex = 0; logCategoryIndex < logLevelList.Count; logCategoryIndex++)
+                    {
+                        SetLogLevel((LogCategory)logCategoryIndex, logLevelList[logCategoryIndex]);
+                    }
 
                     InitializeOverlay(coroutineOwner);
                     return;
                 }
 
 #if !UNITY_EDITOR && !UNITY_SWITCH
                 // Set logging to VeryVerbose on EOS SDK bootstrap so we get the most logging information
-                SetLogLevel(LogCategory.AllCategories, LogLevel.VeryVerbose);
+                for (int logCategoryIndex = 0; logCategoryIndex < logLevelList.Count; logCategoryIndex++)
+                {
+                    SetLogLevel((LogCategory)logCategoryIndex, logLevelList[logCategoryIndex]);
+                }
```
It's unclear to me why this loop is the same in both places here - does it really need to be repeated?

arthur740212 replied:
They are ran under different conditions. One for running in the Editor, the other for Player.

paulhazen commented on Assets/Plugins/Source/Core/EOSManager.cs:
```diff
@@ -675,12 +680,10 @@ public void Init(IEOSCoroutineOwner coroutineOwner, string configFileName)
-#else
-                SetLogLevel(LogCategory.AllCategories, LogLevel.Warning);
-#endif
+                for (int logCategoryIndex = 0; logCategoryIndex < logLevelList.Count; logCategoryIndex++)
+                {
+                    SetLogLevel((LogCategory)logCategoryIndex, logLevelList[logCategoryIndex]);
+                }
```
Same question here as above; it does not seem like between each of these loops the log level is changing, so I'm confused about why each needs to be run.

arthur740212 replied:
These are for the platforms that uses dynamic binding, which are excluded in the earlier call.
Those using dynamic binding can not use SetLogLevel before the LogInterface gets initialized, or else it would crash.

paulhazen commented on Assets/Plugins/Source/Core/LogLevelConfig.cs:
```diff
@@ -0,0 +1,48 @@
+        public string category;
+        public string level;
+
+        public LogCategoryLevelPair(string _category, string _level)
+        {
+            category = _category;
+            level = _level;
+        }
```
I don't think this is a naming convention I have seen before. I would suggest having the parameters named the same, but omit the preceding `_`.

To differentiate between the parameters and the member variables you could either add `this.` before each assignment, or you could make the member variables capitalized (probably the better way to do it - since they are public members).

andrew-hirata-playeveryware replied:
I have seen a similar naming convention before, but only for the other way around: a leading underscore being used to mark a private variable of a class. I would prefer changing the parameter names to something like `aCategory` and `aLevel` , or using `this` to disambiguate the code.

arthur740212 replied:
Fixed!

paulhazen commented on Assets/Plugins/Source/Core/LogLevelConfig.cs:
```diff
@@ -0,0 +1,48 @@
+    }
+
+    [Serializable]
+    public class LogLevelConfig : Config
```
Is `LogLevelConfig` only ever utilized within the editor? If so, it should extend from `EditorConfig`.

arthur740212 replied:
The log window in the samples will use them in runtime eventually.

paulhazen commented on Assets/Plugins/Source/Core/LogLevelHelper.cs:
```diff
@@ -0,0 +1,86 @@
+            get { return Enum.GetNames(typeof(LogLevel)); }
+        }
+
+        public static LogLevelConfig LoadLogLevelConfigFileFromPath(string logLevelConfigPath)
+        {
+            string configDataAsString = "";
+#if UNITY_ANDROID && !UNITY_EDITOR
+            configDataAsString = AndroidFileIOHelper.ReadAllText(logLevelConfigPath);
+#else
+
+            if (!File.Exists(logLevelConfigPath))
+            {
+                throw new Exception("Couldn't find Log Level Config file: Please ensure log_level_config.json exists and is a valid config");
+            }
+
+            configDataAsString = File.ReadAllText(logLevelConfigPath);
+#endif
+            var configData = JsonUtility.FromJson<LogLevelConfig>(configDataAsString);
+
+            Debug.Log("Loaded log level config file: " + configDataAsString);
+            return configData;
+        }
```
This function is no longer needed since we have implemented the `Config.Get<T>()` method for all classes that derive from `Config`. See my next comment for what to do so that this whole function can be deleted.

arthur740212 replied:
I would mostly agree, but the `Config.Get<T>()` function does not handle file reading on Android.

paulhazen commented on Assets/Plugins/Source/Core/LogLevelHelper.cs:
```diff
@@ -0,0 +1,86 @@
+        {
+            get
+            {
+                LogLevelConfig logLevelConfig = LoadLogLevelConfigFileFromPath(Path.Combine(Application.streamingAssetsPath, "EOS", "log_level_config.json"));
```
This should be replaced with the following:

```c#
LogLevelConfig logLevelConfig = Config.Get<LogLevelConfig>();
```

Implementing the `Get` function was done so that we don't have to repeat the code written in the function `LoadLogLevelConfigFileFromPath` (see earlier comment).

arthur740212 replied:
Used `JsonUtility.FromJson<>()` instead, which handles Android file reading

paulhazen reviewed (changes requested):
I think there are a few areas that need some attention. Additionally, I would suggest that instead of having a dropdown for each section that allows us to select what log level to use, have a set of lists - one for each log level, and within each list you can have the category, then the user can drag the category to whichever level they want. This would make the UI a lot more intuitive, and far less noisy.

arthur740212 committed:
chore(style) : disambiguate variables

arthur740212 committed:
chore(json) : use simplified function to read/parse json

arthur740212 committed:
fix(build) : Add EOS_DISABLE on log config scripts

paulhazen commented on Assets/Plugins/Source/Editor/Windows/LogLevelWindow.cs:
```diff
@@ -0,0 +1,118 @@
+    {
+        LogLevelConfig currentLogLevelConfig;
+
+        string[] categories = LogLevelHelper.LogCategoryStringArray;
+        string[] levels = LogLevelHelper.LogLevelStringArray;
```
I don't think these should be stored as member variables - they are only ever needed once in the lifecycle of the window, so you should be able to query them when you need them.

arthur740212 replied:
It is used a couple of times in the script.
I think it is cleaner to shorthand these.

paulhazen replied:
Fair enough :)

paulhazen commented on Assets/Plugins/Source/Editor/Windows/LogLevelWindow.cs:
```diff
@@ -0,0 +1,118 @@
+                    currentLogLevelConfig.logCategoryLevelPairs.Add(new LogCategoryLevelPair(category, "Info"));
+                }
+
+                SaveToJSONConfig(true);
```
Why are you saving the JSON inside the `Setup` function?

arthur740212 replied:
This is under the if statement when the list of categories has changed (for scenarios like a new interface from SDK),
and it regenerates a new config file in such cases.

paulhazen replied:
I see - so basically - when the Window is opened, it loads the current log level configuration - then checks to see if there are any log levels that aren't included in it - and if there are, it updates them accordingly and saves the JSON. 

Am I understanding that correctly?

arthur740212 replied:
Partially. It creates a new config and set them all to LogLevel.Info.
The usage of this is pretty rare, since EOSSDK wouldn't dish out new categories that often.

paulhazen commented on Assets/Plugins/Source/Editor/Windows/LogLevelWindow.cs:
```diff
@@ -0,0 +1,118 @@
+
+
+
+            if (GUILayout.Button("Save All Changes"))
+            {
+                SaveToJSONConfig(true);
+            }
+        }
+
+        private void SaveToJSONConfig(bool prettyPrint)
+        {
+            currentLogLevelConfig.Write(prettyPrint);
+        }
```
It feels like this can be simplified. You can remove the function `SaveToJSONConfig` (functions with that name have been removed from all the other windows that used to utilize it). Subsequent to removing that function, you can replace line `108` with simply `currentLogLevelConfig.Write()` (the `prettyPrint` parameter defaults to `true` if I recall correctly).

paulhazen commented on Assets/StreamingAssets/EOS/log_level_config.json:
Can we add some documentation that makes it clear what the delineation is between each of these categories? It feels like it may be prone to arbitrary assignment. Is it possible that instead we might want to categorize the logs based on the scene?

arthur740212 replied:
These categories are all grabbed directly from EOSSDK log category enums, I'm not sure what you mean.

paulhazen replied:
Ah - I'm sorry I misunderstood - I thought that _we_ were defining these categories. I'll mark this as resolved.

paulhazen commented on Assets/Plugins/Source/Core/LogLevelConfig.cs:
```diff
@@ -0,0 +1,48 @@
+namespace PlayEveryWare.EpicOnlineServices
+{
+    [Serializable]
+    public class LogCategoryLevelPair
+    {
+        public string category;
+        public string level;
+
+        public LogCategoryLevelPair(string category, string level)
+        {
+            this.category = category;
+            this.level = level;
+        }
+    }
+
+    [Serializable]
+    public class LogLevelConfig : Config
```
Because these two classes are only ever utilized by the `LogLevelHelper` class - I suggest that they be made into classes that are nested with in. Furthermore, if I'm not mistaken the first class can also be made to be nested within the second one.

WispyMouse replied:
This suggestion makes sense to me. It's really limited in its use cases, and would add fewer things to the namespace for people to be concerned with.

arthur740212 replied:
The reason why they were separated is because the Helper class interacts with the SDK, while the Config only handles pure text/json without knowing the SDK exists

paulhazen replied:
Your observation that the Utility class interacts with the SDK while the Config only handles pure text/json is an astute one. I think your reasoning is salient. However - I would argue that making these classes nested does not actually make them interact with the SDK at all - it just limits the scope of where those classes can be utilized.

This is _very_, _very_ pedantic of me (and I take no issue with your reasoning here) but thought I would give voice to my thought-process in case it facilitates more thought on the topic in future circumstances that may be different.

paulhazen commented on Assets/Plugins/Source/Core/LogLevelHelper.cs:
```diff
@@ -0,0 +1,68 @@
+
+namespace PlayEveryWare.EpicOnlineServices
+{
+    public static class LogLevelHelper
```
For most classes of this ilk, we have been using the nomenclature of "Utility," in favor of "Helper." To maintain consistency within the project I recommend this particular class be renamed `LogLevelUtility`, and be moved into the `Source/Core/Utility` directory.

paulhazen reviewed (changes requested):
Please see comments for details of my requested changes. Additionally, I believe that we need to put this PR through the prospectus process.

WispyMouse commented on Assets/Plugins/Source/Core/EOSManager.cs:
```diff
@@ -584,19 +586,22 @@ public void Init(IEOSCoroutineOwner coroutineOwner, string configFileName)
-                    SetLogLevel(LogCategory.AllCategories, LogLevel.Warning);
-#endif
+
+                    for (int logCategoryIndex = 0; logCategoryIndex < logLevelList.Count; logCategoryIndex++)
```
I feel like since this shows up in three places, we should make a private function that its purpose is to do this code.

WispyMouse commented on Assets/Plugins/Source/Core/EOSManager.cs:
```diff
@@ -584,19 +586,22 @@ public void Init(IEOSCoroutineOwner coroutineOwner, string configFileName)
                 }
 
 #if !UNITY_EDITOR && !UNITY_SWITCH
                 // Set logging to VeryVerbose on EOS SDK bootstrap so we get the most logging information
```
Comment can be nixed, no longer true

WispyMouse commented on Assets/Plugins/Source/Core/LogLevelConfig.cs:
```diff
@@ -0,0 +1,48 @@
+        public string category;
+        public string level;
+
+        public LogCategoryLevelPair(string category, string level)
```
My instincts tell me that if a class is Serializable, it needs to have a parameterless constructor if it's going to be serialized/deserialized. That may or may not be true here, but I'd like to just check in that this is deserializing properly.

arthur740212 replied:
The serialize attribute is for the two member variables to work with json. 
The constructor is only called from the window and probably has nothing to do with json.

Update : I think you are worried if this will break in the inspector?
We don't use built in inspector features for this config, so we don't need to worry about that part.

WispyMouse commented on Assets/Plugins/Source/Core/LogLevelConfig.cs:
```diff
@@ -0,0 +1,48 @@
+    }
+
+    [Serializable]
+    public class LogLevelConfig : Config
```
This needs to have a static constructor that registers it to the factory. When trying to use this I get;

> No factory method has been registered for type "PlayEveryWare.EpicOnlineServices.LogLevelConfig". Please make sure that "PlayEveryWare.EpicOnlineServices.LogLevelConfig" registers its constructor with the base Config class via a static constructor.

arthur740212 committed:
chore(rename) : Rename and relocate LogLevelUtility

WispyMouse committed:
fix (LogLevelConfig) : Register factory

arthur740212 committed:
chore(comment) : remove outdated comment

arthur740212 committed:
chore(optimize) : make repeated code into function InitializeLogLevels

arthur740212 committed:
chore(optimize) : simplify code for saving config

arthur740212 committed:
fix(editor,log) : register factory with static constructor

RuntimeHelpers.RunClassConstructor is used in TryGetFactory to explicitly call the static constructor of the Config.
This constructor must be changed to static or else the instance constructor will not be registered

arthur740212 committed:
chore(comment) : comments to describe setup code

paulhazen commented on Assets/Plugins/Source/Editor/Windows/LogLevelWindow.cs:
```diff
@@ -0,0 +1,114 @@
+        {
+            currentLogLevelConfig = EpicOnlineServices.Config.Get<LogLevelConfig>();
+
+            // Config.Get<T> only guarantees T to be not null, but not for its members
```
Excellent observation :thinking: might have to consider this fact if ever there are future adjustments.

paulhazen commented on Assets/Plugins/Source/Core/EOSManager.cs:
```diff
@@ -769,6 +759,20 @@ public void SetLogLevel(LogCategory Category, LogLevel Level)
 
+            //-------------------------------------------------------------------------
+            /// <summary>
+            /// Initialize log levels loaded from <see cref="LogLevelConfig" />
```
+100 for using cool `cref` comment linkage. Makes me want to up my own commenting game 😃

paulhazen reviewed (changes requested):
The most recent updates to this branch are baller! Kudos for continuing to grind on this, it's great work, and I hope you're okay with us all putting it through the proverbial wringer :)

My only remaining concern is that I do not see entries in `eos_package_description.json` for the files that have been added. If I'm not mistaken - this functionality would be good to have included in the package when imported from UPM.

paulhazen committed:
Merge branch 'development' into log_config_redo

arthur740212 committed:
feat(log,native) : Set log level with config for native libs

andrew-hirata-playeveryware commented on Assets/Plugins/Source/Core/EOSManager.cs:
```diff
@@ -673,13 +669,7 @@ public void Init(IEOSCoroutineOwner coroutineOwner, string configFileName)
-
-                // Default back to quiet logs
-#if UNITY_EDITOR
-                SetLogLevel(LogCategory.AllCategories, LogLevel.VeryVerbose);
```
I think in editor we should still use 'very verbose'

arthur740212 committed:
fix(log) : default log level veryverbose on editor

arthur740212 committed:
fix(log) : set log level for initialize depending on library link type on platform

Set log level for initialize depending on library link type on platform
Removed set log level post initialize because it doesn't reflect which level the app was using.

andrew-hirata-playeveryware commented on Assets/Plugins/Source/Core/EOSManager.cs:
```diff
@@ -1772,8 +1771,8 @@ private static void UpdateNetworkStatus()
-        /// <value>Private static instance of <c>EOSSingleton</c></value>
-        static EOSSingleton s_instance;
+                /// <value>Private static instance of <c>EOSSingleton</c></value>
+                static EOSSingleton s_instance;
```
Remove the extra tabs, if you could.

andrew-hirata-playeveryware commented on Assets/Plugins/Source/Core/EOSManager.cs:
```diff
@@ -584,23 +584,23 @@ public void Init(IEOSCoroutineOwner coroutineOwner, string configFileName)
 
+                // Set log level prior to platform interface initialization
+                // VeryVerbose for dynamic linking platforms, otherwise set levels from configs 
+#if UNITY_EDITOR || UNITY_SWITCH
```
I believe you can remove the extra UNITY_SWITCH define.

arthur740212 committed:
fix(log) : remove extra tabs and ifdefs

paulhazen commented on Assets/Plugins/Source/Editor/Windows/LogLevelWindow.cs:
```diff
@@ -0,0 +1,114 @@
+        bool showAllCategories = false;
+        string previousAllCategoryLogLevel;
+
+        [MenuItem("Tools/EOS Plugin/Log Level")]
+        public static void ShowWindow()
+        {
+            GetWindow<LogLevelWindow>("Log Level");
+        }
```
For consistency, I think we should rename the window and the corresponding MenuItem `Tools/EOS Plugin/Log Level Configuration`.

paulhazen commented on Assets/Plugins/Source/Editor/Windows/LogLevelWindow.cs:
```diff
@@ -0,0 +1,114 @@
+                currentLogLevelConfig.logCategoryLevelPairs = new List<LogCategoryLevelPair>();
+            }
+
+            // Initialize the list if config categories does not match the SDK categories, which might happen when using a different SDK version
+            if (currentLogLevelConfig.logCategoryLevelPairs.Count != categories.Length)
```
This will not catch circumstances where a category is renamed, or if between one SDK and another there is one log category removed and one log category added.

arthur740212 replied:
That is true, but will be ultra rare, since it only happens with a change from the SDK.
The category names are only used when creating the json, and does not check the names when reading,
so rename won't be a problem.
removing and adding a category is even more rare, but in those case when it happens, the categories and levels are still still represented correctly in the UI.

The purpose of this is to make sure there are not any out of range exceptions

paulhazen replied:
That's a good point regarding the intent of the check. It does however reveal that the aforementioned circumstance can happen and is not accounted for by the current implementation. If this circumstance _did_ happen (and I completely agree it would be extraordinarily rare) I can see it causing a fair amount of confusion and subsequent mayhem.

paulhazen commented on Assets/Plugins/Source/Editor/Windows/LogLevelWindow.cs:
```diff
@@ -0,0 +1,114 @@
+        protected override void RenderWindow()
+        {
+            // Last element in the list is All Categories
+            previousAllCategoryLogLevel = currentLogLevelConfig.logCategoryLevelPairs[categories.Length - 1].level;
```
I think this line should be moved to be between line 97 and line 98. More importantly, it should be a local variable - there is no need to store it as a member of the `LogLevelWindow` class, as it is only ever used in this circumstance - and its value does not need to persist between calls of `RenderWindow`.

arthur740212 replied:
You are right, it should be local.
But it cannot be moved, because the level change should be between previous and current to check if it is dirty.

paulhazen commented on Assets/Plugins/Source/Core/LogLevelConfig.cs:
```diff
@@ -0,0 +1,52 @@
+    [Serializable]
+    public class LogCategoryLevelPair
+    {
+        public string category;
+        public string level;
```
These should be capitalized.

paulhazen commented on Assets/Plugins/Source/Core/LogLevelConfig.cs:
```diff
@@ -0,0 +1,52 @@
+        }
+        public LogLevelConfig() : base("log_level_config.json") { }
+
+        public List<LogCategoryLevelPair> logCategoryLevelPairs;
```
This should also be capitalized.

paulhazen commented on Assets/Plugins/Source/Core/Utility/LogLevelUtility.cs:
```diff
@@ -0,0 +1,68 @@
+        {
+            get
+            {
+                LogLevelConfig logLevelConfig = JsonUtility.FromJsonFile<LogLevelConfig>(Path.Combine(Application.streamingAssetsPath, "EOS", "log_level_config.json"));
```
This should be changed to just: `Config.Get<LogLevelConfig>()`

arthur740212 replied:
Is the Android IO removed?
This was made like this because of that

arthur740212 replied:
The IO fix will be on another branch anyways, so yeah, will change this.

paulhazen commented on Assets/Plugins/Source/Core/Utility/LogLevelUtility.cs:
```diff
@@ -0,0 +1,68 @@
+            {
+                LogLevelConfig logLevelConfig = JsonUtility.FromJsonFile<LogLevelConfig>(Path.Combine(Application.streamingAssetsPath, "EOS", "log_level_config.json"));
+
+                var logLevels = new List<LogLevel>();
```
Instead of having the ambiguity at the beginning of the variable declaration in the form of the `var` keyword, a better approach would be to utilize the `new()` constructor like this:

```csharp
List<LogLevel> logLevels = new()
```

paulhazen commented on Assets/Plugins/Source/Editor/Windows/LogLevelWindow.cs:
```diff
@@ -0,0 +1,114 @@
+{
+    public class LogLevelWindow : EOSEditorWindow
+    {
+        LogLevelConfig currentLogLevelConfig;
```
There should be no need to store a copy of the log level config. If the log levels change while this is running, there is a chance that the wrong values could be subsequently saved. 

It is safer to use: 
```csharp
Config.Get<LogLevelConfig>();
```

arthur740212 replied:
If it is not stored somewhere, then it will need to reload with `Config.Get` every frame the window renders, which will make the config immutable.

arthur740212 committed:
chore(format) : Capitalizations, name changes, variable scope changes

arthur740212 committed:
fix(log) : Make sure the config always uses the right categories
